wandb/sdk/launch/docker.py

```diff
@@ -64,139 +51,151 @@ def generate_docker_image(
+        data = docker.run(["docker", "image", "inspect", docker_image])
+        parsed = json.loads(data)[0]
+        _inspected_images[docker_image] = parsed
+        return True
+    except (DockerError, ValueError) as e:
+        if should_raise:
+            raise e
+        return False
 
 
-def build_docker_image(
-    launch_project: _project_spec.LaunchProject, base_image: str, copy_code: bool,
-) -> Union[Model, Any]:
-    """Build a docker image containing the project in `work_dir`, using the base image.
+def docker_image_inspect(docker_image: str) -> Dict[str, Any]:
+    """Get the parsed json result of docker inspect image_name"""
+    if _inspected_images.get(docker_image) is None:
+        docker_image_exists(docker_image, True)
+    return _inspected_images.get(docker_image, {})
 
-    Arguments:
-    launch_project: LaunchProject class instance
-    base_image: base_image to build the docker image off of
-    api: instance of wandb.apis.internal Api
-    copy_code: boolean indicating if code should be copied into the docker container
 
-    Returns:
-        A `Model` instance of the docker image.
+def pull_docker_image(docker_image: str) -> None:
+    """Pulls the requested docker image"""
+    try:
+        docker.run(["docker", "pull", docker_image])
+    except DockerError as e:
+        raise LaunchError("Docker server returned error: {}".format(e))
+
 
-    Raises:
-        LaunchError: if there is an issue communicating with the docker client
+def build_docker_image_if_needed(
+    launch_project: _project_spec.LaunchProject, api: Api, copy_code: bool
+) -> str:
+    """
+    Build a docker image containing the project in `work_dir`, using the base image.
+    param launch_project: LaunchProject class instance
+    :param api: instance of wandb.apis.internal Api
+    :param copy_code: boolean indicating if code should be copied into the docker container
     """
-    import docker
 
-    image_name = "wandb_launch_{}".format(launch_project.run_id)
     image_uri = _get_docker_image_uri(
-        name=image_name, work_dir=launch_project.project_dir
+        name=launch_project.image_name, work_dir=launch_project.project_dir
     )
+    launch_project.docker_image = image_uri
+    if docker_image_exists(image_uri):
+        wandb.termlog("Using existing image: {}".format(image_uri))
+        return image_uri
+    container_inspect = docker_image_inspect(launch_project.base_image)
     copy_code_line = ""
-    workdir_line = ""
-    copy_config_line = ""
-    workdir = os.path.join("/home/", getpass.getuser())
-    if launch_project.override_config:
-        copy_config_line = "COPY {}/{} {}\n".format(
-            _PROJECT_TAR_ARCHIVE_NAME, _project_spec.DEFAULT_CONFIG_PATH, workdir
-        )
+    requirements_line = ""
+    workdir = container_inspect["ContainerConfig"].get("WorkingDir", "/")
+    # TODO: we currently assume the home directory holds the pip cache
```

Comment:
This is a little brittle for custom base_images.  There's probably a better way to introspect the pip cache location, we just assume it's in `~/.cache`